### ISSUE TITLE
STIX Patterning needs to support basic type operations

### ISSUE THREAD

**Issue by treyka** [oasis-tcs/cti-stix2#62]
|      function | description                                     |
|----------------------------------------------------|---------------------------------------------------------------------------------------------------|
| Casting (or, define casting behaviour in the spec) | base64 => hex, utf8 string, int, float, etc (Plus the combinatorial of the above for other types) |

**Comment by gtback** [comment]
@treyka do you have an example of when this would be useful, compared to just ensuring the values in the pattern are the desired format (and maybe some normative text explaining how pattern evaluation tools should handle mismatched types)?

**Comment by jmgnc** [comment]
@gtback For example, there are some properties which are stored in binary format, but are UTF-8 string, so being able to do: `utf8(property_bin) MATCHES 'someregexwithutf8characters'`  is much better and more readable.

For int/float, I expect those to be common, as there are often times people will be lazy and return a number property as a string instead of an integer requiring them.  Once we have the ability to extract substring matches, or string manipulation, then converting parts of substrings to numbers will be more common.

**Comment by gtback** [comment]
Every property of Cyber Observables has a defined data type. If the source of data you're using for evaluating patterns isn't already the correct type, it needs to be converted to that type before a pattern can be evaluated. That's not something patterns should need to account for.

If the issue is "I want to check a binary property, but I know the data in that field will really be UTF-8 text, and it's easier/more readable for the pattern to be text than hex/base64", I think that's more of an issue for the tool that generates the pattern than something we want to support in the pattern language itself.

**Comment by jmgnc** [comment]
Until we have such a tool to "invert checks", we should add it..  yes, you can convert `int(x) < 182379454 and int(x) > 8347485` into a regex when x is a string, but I definitely don't want to do that by hand, and I don't know of a tool that will do that currently.  But what if x is sometimes decimal and other times hexadecimal?   it'd further compound the difficulty.

**Comment by gtback** [comment]
Do you have a specific `x` where this would be used? I'm not saying there's no use case (hence my question from January), but I absolutely want to stay away from making pattern matching harder to implement based on "it would be cool if...".

**Comment by jmgnc** [comment]
I do not have a specific x yet for this use case, assuming that new SCOs are approved in a timely fashion (which may be an issue w/ the new rules).  If SCO's aren't standardized that do the decoding of necessary fields, then the only way to write portable patterns is to use one of the two methods out lined above.

There are a number of other features I'd much rather see before this.

**Comment by jmgnc** [comment]
Oh, going back over things, I do have a use case for decoding base64 (and other string functions).  Currently the http request SCO extension does not break out the authorization header when it's basic.  If you need to check username, or password, the regex to detect this is complicated as you can only check the encode base64, so you have to have 3 different expressions for each "group" of base64 characters, and if you're doing a more complicated check than a simple equality (i.e. username MATCHES (bob|mary|sam) ), then that grows it much larger.

**Comment by gtback** [comment]
As I understand it, HTTP Basic authorization doesn't base64-encode the entire value of the `Authorization` header, just the `username:password', meaning you'd likely need to substring the part after `Basic ` anyway before decoding (and decoding is not quite the same thing as casting anyway unless you have a "cast" like `b64decode()`).

More generally, though, **I do see that there *could* be the use for this, particularly in `dictionary` properties, where things are "stringly-typed"**. In this example, though, if there *were* a common case in STIX for needing to match the username/password in an HTTP Basic Authorization header, it might be better to extract that into true Cyber Observable properties:

`[network-traffic:extensions."http-request-ext"."basic-authorization".username IN (bob, mary, sam)]` 

rather than

`[b64decode(substring(network-traffic:extensions."http-request-ext"."request-header".Authentication, 6)) MATCHES '^(bob|mary|sam)']` (or something like that.)

**Comment by jmgnc** [comment]
Correct, which is why I included other string functions as needed to implement this use case.

The issue is that this isn't the only HTTP header, such a cookies that will be base64 encoding.

We need to balance between decoding everything in the spec, and the patterning language doing the work.  The other issue is that this will depend upon the SCO generator populating these fields.  It's much more likely that the header fields will be populated, but it's less likely that these other fields will be.  As the spec marks most fields optional, there is no requirement that if there is an authorization header present that it be properly decoded.

This is an issue that I've been thinking about since I started working on this.  And it's not easy to decide where the line should be drawn.  Even harder when you have a committee involved.

**Comment by jordan2175** [comment]
We talked about this on 2019-06-05 and Jason agreed to propose the text in the document. 

We think we should start small and do the basics for 2.1.
Cast numbers to strings and strings to numbers
Cast to and from strings
